File: scripts/extract-translations.py
#!/usr/bin/env python3
"""
Extract translations from dictionaries.ts and create modular JSON files.
"""
import json
import re
import os
from pathlib import Path

# Read the dictionaries.ts file
dict_path = Path(__file__).parent.parent / 'src' / 'lib' / 'i18n' / 'dictionaries.ts'
with open(dict_path, 'r', encoding='utf-8') as f:
    content = f.read()

# Extract language blocks
def extract_lang_block(lang_code):
    """Extract a language block from the TypeScript file."""
    pattern = rf'const {lang_code}: Dictionary = \{{(.*?)\}};'
    match = re.search(pattern, content, re.DOTALL)
    if not match:
        return None
    return match.group(1)

# Simple TS to JSON converter (handles basic cases)
def ts_value_to_json(value_str):
    """Convert TypeScript value to JSON value."""
    value_str = value_str.strip()
    
    # Remove comments
    value_str = re.sub(r'//.*?$', '', value_str, flags=re.MULTILINE)
    value_str = re.sub(r'/\*.*?\*/', '', value_str, flags=re.DOTALL)
    
    # String
    if value_str.startswith('"') or value_str.startswith("'"):
        return value_str.strip('"\'')
    
    # Number
    if value_str.replace('.', '').replace('-', '').isdigit():
        return float(value_str) if '.' in value_str else int(value_str)
    
    # Boolean
    if value_str in ('true', 'false'):
        return value_str == 'true'
    
    # null
    if value_str == 'null':
        return None
    
    # Object (simplified)
    if value_str.startswith('{'):
        return parse_ts_object(value_str)
    
    # Array (simplified)
    if value_str.startswith('['):
        return parse_ts_array(value_str)
    
    return value_str

def parse_ts_object(obj_str):
    """Parse a TypeScript object to Python dict."""
    result = {}
    # This is a simplified parser - for production, use a proper TS parser
    # Remove outer braces
    obj_str = obj_str.strip('{}')
    
    # Split by commas, but respect nested objects/arrays
    depth = 0
    in_string = False
    escape_next = False
    current_key = None
    current_value = ''
    start_idx = 0
    
    i = 0
    while i < len(obj_str):
        char = obj_str[i]
        
        if escape_next:
            escape_next = False
            i += 1
            continue
        
        if char == '\\':
            escape_next = True
            i += 1
            continue
        
        if char in ('"', "'", '`'):
            in_string = not in_string
            i += 1
            continue
        
        if in_string:
            i += 1
            continue
        
        if char == '{' or char == '[':
            depth += 1
        elif char == '}' or char == ']':
            depth -= 1
        elif char == ':' and depth == 0:
            # Key-value separator
            current_key = obj_str[start_idx:i].strip().strip('"\'')
            start_idx = i + 1
        elif char == ',' and depth == 0:
            # End of current key-value pair
            current_value = obj_str[start_idx:i].strip()
            if current_key:
                result[current_key] = ts_value_to_json(current_value)
            current_key = None
            current_value = ''
            start_idx = i + 1
        
        i += 1
    
    # Handle last pair
    if current_key:
        current_value = obj_str[start_idx:].strip()
        result[current_key] = ts_value_to_json(current_value)
    
    return result

def parse_ts_array(arr_str):
    """Parse a TypeScript array to Python list."""
    # Simplified - just return empty for now
    return []

# Section extraction mappings
def extract_section(data, path):
    """Extract a nested section from the data."""
    keys = path.split('.')
    current = data
    for key in keys:
        if isinstance(current, dict) and key in current:
            current = current[key]
        else:
            return None
    return current

# For now, let's use a manual approach since TS parsing is complex
# We'll create the JSON files based on the structure we know

print("Translation extraction script")
print("Note: Due to TypeScript parsing complexity, JSON files will be created manually")
print("based on the dictionaries.ts structure.")

File: scripts/extract-translations.js
const fs = require('fs');
const path = require('path');

// Read the dictionaries.ts file
const dictPath = path.join(__dirname, '../src/lib/i18n/dictionaries.ts');
const content = fs.readFileSync(dictPath, 'utf-8');

// Extract the translation objects using regex
const extractObject = (lang, startMarker, endMarker) => {
  const start = content.indexOf(`${lang}: Dictionary = {`);
  if (start === -1) return null;
  
  let depth = 0;
  let inString = false;
  let escapeNext = false;
  let startPos = content.indexOf('{', start);
  let endPos = startPos;
  
  for (let i = startPos; i < content.length; i++) {
    const char = content[i];
    
    if (escapeNext) {
      escapeNext = false;
      continue;
    }
    
    if (char === '\\') {
      escapeNext = true;
      continue;
    }
    
    if (char === '"' || char === "'" || char === '`') {
      inString = !inString;
      continue;
    }
    
    if (inString) continue;
    
    if (char === '{') {
      depth++;
    } else if (char === '}') {
      depth--;
      if (depth === 0) {
        endPos = i + 1;
        break;
      }
    }
  }
  
  return content.substring(startPos, endPos);
};

// Simple parser to convert TS object to JSON (handles basic cases)
const tsToJson = (tsStr) => {
  // Remove comments
  tsStr = tsStr.replace(/\/\/.*$/gm, '');
  tsStr = tsStr.replace(/\/\*[\s\S]*?\*\//g, '');
  
  // Convert TypeScript object to JSON-like string
  // This is a simplified parser - handles basic cases
  let jsonStr = tsStr
    .replace(/(\w+):/g, '"$1":') // Convert keys to quoted
    .replace(/,\s*}/g, '}') // Remove trailing commas
    .replace(/,\s*]/g, ']') // Remove trailing commas in arrays
    .replace(/\{(\w+):/g, '{"$1":'); // Handle nested objects
  
  try {
    return JSON.parse(jsonStr);
  } catch (e) {
    // Fallback: manual extraction for complex cases
    return null;
  }
};

// Section mappings
const sections = {
  navbar: 'landing.nav',
  home: ['landing.hero', 'landing.features', 'landing.demo', 'landing.howItWorks', 'landing.about', 'landing.cta', 'landing.story', 'landing.workflow', 'landing.ecosystem', 'landing.trust', 'landing.finalCta'],
  footer: 'landing.footer',
  dashboard: 'admin.dashboard',
  menu: ['menu', 'admin.menus', 'admin.categories'],
  order: 'admin.orders',
  table: {}, // Empty for now
  booking: {}, // Empty for now
  analytics: 'admin.analytics',
  appearance: 'admin.appearance',
  qr: 'admin.qr',
  settings: 'admin.settings',
};

// Languages
const languages = ['en', 'de', 'fr', 'it'];

// For now, let's create a simpler approach: manually extract based on what we know
// This script will be a helper to create the structure

console.log('Translation extraction script');
console.log('Note: This is a helper script. Manual extraction may be needed for complex nested structures.');
